apps/constellation/manifest_builder.cpp

```diff
@@ -0,0 +1,141 @@
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "constants.hpp"
+#include "manifest_builder.hpp"
+
+#include "core/byte_array/const_byte_array.hpp"
+#include "core/byte_array/decoders.hpp"
+#include "core/filesystem/read_file_contents.hpp"
+#include "network/p2pservice/manifest.hpp"
+#include "network/p2pservice/p2p_service_defs.hpp"
+#include "network/peer.hpp"
+#include "settings.hpp"
+
+namespace fetch {
+namespace {
+
+using fetch::byte_array::ConstByteArray;
+using fetch::byte_array::FromBase64;
+using fetch::network::Manifest;
+using fetch::network::Peer;
+using fetch::network::ServiceIdentifier;
+using fetch::network::ServiceType;
+using fetch::network::Uri;
+
+/**
+ * Generate a default simple manifest for all the services provided
+ *
+ * @param external_address The externally visible address to communicate on
+ * @param port The starting port for the services
+ * @param num_lanes The number of lanes to be configured
+ * @param manifest The output manifest to be populated
+ */
+void GenerateDefaultManifest(std::string const &external_address, uint16_t port, uint32_t num_lanes,
+                             Manifest &manifest)
+{
+  Peer peer;
+
+  // register the HTTP service
+  peer.Update(external_address, port + HTTP_PORT_OFFSET);
+  manifest.AddService(ServiceIdentifier{ServiceType::HTTP}, Manifest::Entry{Uri{peer}});
+
+  // register the P2P service
+  peer.Update(external_address, static_cast<uint16_t>(port + P2P_PORT_OFFSET));
+  manifest.AddService(ServiceIdentifier{ServiceType::CORE}, Manifest::Entry{Uri{peer}});
+
+  // register all of the lanes (storage shards)
+  for (uint32_t i = 0; i < num_lanes; ++i)
+  {
+    peer.Update(external_address, static_cast<uint16_t>(port + STORAGE_PORT_OFFSET + i));
+
+    manifest.AddService(ServiceIdentifier{ServiceType::LANE, static_cast<uint16_t>(i)},
+                        Manifest::Entry{Uri{peer}});
+  }
+}
+
+/**
+ * Create the manifest from a specified file path
+ *
+ * @param filename The filename to try and load
+ * @param manifest The output manifest to be populated
+ * @return true if successful, otherwise false
+ */
+bool LoadManifestFromFile(char const *filename, Manifest &manifest)
+{
+  ConstByteArray buffer = core::ReadContentsOfFile(filename);
+
+  // check to see if the read failed
+  if (buffer.size() == 0)
+  {
+    return false;
+  }
+
+  return manifest.Parse(buffer);
+}
+
+/**
+ * Create the manifest from an environment variable based configuration
+ *
+ * @param manifest The manifest to be populated
+ * @return true if successful, otherwise false
+ */
+bool LoadManifestFromEnvironment(Manifest &manifest)
+{
+  bool success{false};
+
+  char const *manifest_data = std::getenv("CONSTELLATION_MANIFEST");
```

Comment:
rather than calling std::getenv directly, I'd like to see a thin wrapper around it which we own and can mock out. Aids testability. Rather than a string, the wrapper could take an enum class argument like ConstellationEnvVar, and would look up the right env var string name based on the arg value. This would also have the benefit of putting all env vars constellation depends on in one place - we only have a couple for now, but it's easy to imagine their number growing over time, and the code getting peppered with magic strings like this one.

Reply:
This is sort of a chicken and egg problem a little bit because it is hard to test the value of an environment variable. 

The enum approach would not work because the whole point of this setting change is to allow dynamically query based on the collection settings.

I will raise an issue to track this and I think that we can tackle this as another PR.

Reply:
I think we misunderstand each other - I didn't mean testing values of env variables, but injecting envvariables in a unit test context to check that they override defaults and are overridden by cli arguments. Raise the issue - it's a change we can make on top of this after it's merged.